core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

```diff
@@ -687,63 +688,37 @@ public void authenticateWithMicrosoftCode() {
+                PENDING_MICROSOFT_AUTHENTICATION.getOrCreateTask(this);
 
-        // Use a future to prevent timeouts as all the authentication is handled sync
-        // This will be changed with the new protocol library.
-        CompletableFuture.supplyAsync(() -> {
-            try {
-                return msaAuthenticationService.getAuthCode();
-            } catch (RequestException e) {
-                throw new CompletionException(e);
-            }
-        }).whenComplete((response, ex) -> {
-            if (ex != null) {
-                ex.printStackTrace();
-                disconnect(ex.toString());
-                return;
-            }
-            LoginEncryptionUtils.buildAndShowMicrosoftCodeWindow(this, response);
-            attemptCodeAuthentication(msaAuthenticationService);
-        });
-    }
-
-    /**
-     * Poll every second to see if the user has successfully signed in
-     */
-    private void attemptCodeAuthentication(MsaAuthenticationService msaAuthenticationService) {
-        if (loggedIn || closed) {
-            return;
-        }
-        CompletableFuture.supplyAsync(() -> {
-            try {
-                msaAuthenticationService.login();
-                GameProfile profile = msaAuthenticationService.getSelectedProfile();
-                if (profile == null) {
-                    // Java account is offline
-                    disconnect(GeyserLocale.getPlayerLocaleString("geyser.network.remote.invalid_account", clientData.getLanguageCode()));
-                    return null;
+        if(!task.getAuthentication().isDone()) {
+            task.getCode().whenComplete((response, ex) -> {
+                if (ex != null) {
+                    geyser.getLogger().error("Failed to get Microsoft auth code", ex);
+                    disconnect(ex.toString());
+                } else {
+                    LoginEncryptionUtils.buildAndShowMicrosoftCodeWindow(this, response);
                 }
-
-                return new MinecraftProtocol(profile, msaAuthenticationService.getAccessToken());
-            } catch (RequestException e) {
-                throw new CompletionException(e);
-            }
-        }).whenComplete((response, ex) -> {
+            });
+        }
+        task.getAuthentication().whenComplete((msaAuthenticationService, ex) -> {
             if (ex != null) {
-                if (!(ex instanceof CompletionException completionException) || !(completionException.getCause() instanceof AuthPendingException)) {
-                    geyser.getLogger().error("Failed to log in with Microsoft code!", ex);
-                    disconnect(ex.toString());
+                geyser.getLogger().error("Failed to log in with Microsoft code!", ex);
+                disconnect(ex.toString());
+            } else if (!closed) {
+                GameProfile selectedProfile = msaAuthenticationService.getSelectedProfile();
+                if(selectedProfile == null) {
```

Comment:
lack of space between the if and the condition

Reply:
I missed some other spaces. Noted. Thanks!